Configurable compass drivers based on jflyper's PR #3613

MJ666 opened the pull request:
Based on jflyper's PR #3613.

These drivers are successfully tested with the following boards:

SPRacing F3 Mini (MPU9250 I2C, AK8975 and AK8963 driver with target change)
AlienFlightNG F7 Aubis16 (ICM20602, AK8963 both SPI)
AlienFlightNG F7 Quad Brushed board and my F7 Development Board (MPU2950 SPI)

The MW32 (Naze compatible) (no Mag is included anymore in the NAZE target. I was not able to create a working build with reduced functionality and the Mag added back in. The FC was not responding anymore to the BF Configurator)

This means the AK8963 and AK8975 part is already well tested. The HMC8583L (I2C and SPI) drivers still need to be tested. I don't have any hardware for this part of the tests.

The AlienFlightNG F7 target has support for an MCU9250 (SPI) or the AK8963 (SPI) with another Gyro chip (SPI) in the same target. Both combinations are seamlessly autodetected. The Gyro and Mag sensors are connected to different SPI busses. The SPI Mag is sharing the SPI bus with the Baro without issues (AlienFlightNG F7 Anubis16).

jflyper commented:
Does this properly includes #3613 ?

MJ666 commented:
@jflyper Yes, it included everything from PR #3613 with additional updates to make the Mag drivers work with 3.2. I did not take over the changes for the NUCLEOF7 target since this was likely changed only for testing? Unfortunately rebasing the code from your PR was not going to well so I decided to migrate manually after a few failed attempts.

martinbudden commented on src/main/drivers/compass/compass_ak8975.c:
```diff
@@ -59,29 +62,31 @@
 #define AK8975A_ASAZ 0x12 // Fuse ROM z-axis sensitivity adjustment value
 
-static bool ak8975Init(void)
+static bool ak8975Init(magDev_t *magdev)
```
Why not make this `static bool ak8975Init(busDevice_t *busDev)` since nothing in `magDev_t` seems to be used.

MJ666 replied:
Mhh, this is more consistent with the usage of the other sensor functions? We can change if you prefer.

martinbudden replied:
It's consistent with the others, but I think the others are wrong...
I made a start on changing the other drivers as part of the dual gyro support, but this work is incomplete. So mag be implemented to use `busDevice_t *` and the others will eventually do so as well.

And definitely NO local static pointer - one of the things we want to achieve is making all the drivers re-entrant.

martinbudden commented on src/main/drivers/compass/compass_ak8975.c:
```diff
@@ -90,20 +95,22 @@ static bool ak8975Init(void)
 #define BIT_STATUS2_REG_MAG_SENSOR_OVERFLOW     (1 << 3)
 
-static bool ak8975Read(int16_t *magData)
+static bool ak8975Read(magDev_t *magdev, int16_t *magData)
```
`static bool  ak8975Read(busDevice_t *busDev, int16_t *magData)`

MJ666 replied:
Same as above.

martinbudden commented on src/main/drivers/bus.h:
```diff
@@ -35,13 +35,15 @@ typedef struct busDevice_s {
         struct deviceI2C_s {
            I2CDevice device;
            uint8_t address;
+           const struct busDevice_s *master;
```
Make `busDevice_s` first item in struct for better alignment.

jflyper commented:
@martinbudden Are we happy with the master busdev approach?

martinbudden commented:
I hadn't spotted that. Hmm, it's a bit odd. I won't say I'm happy with it, but then I won't say I'm unhappy either.

MJ666 commented:
This I not to elegant but it works well and I also did not come to an better idea when I was working on this problem before by my self.

MJ666 commented:
Here is my proposal how to change. We even can go even more deep and eliminate the busdev pointer passing of all driver local functions and use the static pointer here instead? Similar as it was in the former version of the drivers. This will have the most impact for the AK8963 driver.

martinbudden commented:
@MJ666 , one of my goals is to eliminate all `static` local variables from the drivers, so they are entirely re-entrant. This was done for gyros to support dual gyros, but in principle we could have dual compasses as well (might be useful for iNav which sort of shares the code).

ledvinap commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -255,48 +343,114 @@ static bool ak8963Read(int16_t *magData)
+    ak8963WriteRegister(busdev, AK8963_MAG_REG_CNTL1, CNTL1_MODE_FUSE_ROM);         // Enter Fuse ROM access mode
+    ak8963ReadRegisterBuffer(busdev, AK8963_MAG_REG_ASAX, sizeof(calibration), calibration); // Read the x-, y-, and z-axis calibration values
+
+    magGain[X] = ((((float)(int8_t)calibration[X] - 128) / 256) + 1) * 30;
```
IMO calibration should be unsigned, it makes little sense with signed.
https://github.com/betaflight/betaflight/pull/3215#pullrequestreview-46061008

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -118,6 +120,7 @@
 #define HMC_POS_BIAS 1
 #define HMC_NEG_BIAS 2
 
+static busDevice_t *busdev;
```
`busdev` shall be passed as argument instead of static

MJ666 replied:
Done.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -118,6 +120,7 @@
 #define HMC_POS_BIAS 1
```
`#define HMC_CONFA_POS_BIAS  0x01`

MJ666 replied:
Done.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -194,31 +264,25 @@ static bool hmc5883lInit(void)
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFA, 0x010 + HMC_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
-#endif
+
+    hmc5883lWriteRegister(HMC58X3_R_CONFA, 0x010 + HMC_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
```
`0x10 | HMC_POS_BIAS`
Also
`#define HMC_CONFA_ODR_15HZ 0x10` and `HMC_CONFA_ODR_15HZ | HMC_CONFA_POS_BIAS`

MJ666 replied:
Removed all magic numbers with proper definitions.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -194,31 +264,25 @@ static bool hmc5883lInit(void)
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFB, 0x60); // Set the Gain to 2.5Ga (7:5->011)
-#endif
+
+    hmc5883lWriteRegister(HMC58X3_R_CONFB, 0x60); // Set the Gain to 2.5Ga (7:5->011)
```
(`#define`)

MJ666 replied:
Done.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -232,21 +296,16 @@ static bool hmc5883lInit(void)
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFA, 0x010 + HMC_NEG_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to negative bias.
-#endif
+
+    hmc5883lWriteRegister(HMC58X3_R_CONFA, 0x010 + HMC_NEG_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to negative bias.
```
(see above)

MJ666 replied:
Done.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -232,21 +296,16 @@ static bool hmc5883lInit(void)
-        i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_MODE, 1);
-#endif
+        hmc5883lWriteRegister(HMC58X3_R_MODE, 1);
         delay(50);
```
Measurement should be ready in 6ms (typ). 20 x 50ms delay is 1s, non-negligible initialization time ...

MJ666 replied:
Reduce to 20 for now. We may go down to 10 if someone can do testing.

ledvinap commented on src/main/sensors/compass.c:
```diff
@@ -54,35 +58,114 @@ mag_t mag;                   // mag access functions
-    .mag_declination = 0,
-    .interruptTag = COMPASS_INTERRUPT_TAG
-);
+void pgResetFn_compassConfig(compassConfig_t *compassConfig)
```
It is possible to `#define` initialization values and then use them in `PG_RESET_TEMPLATE`

MJ666 replied:
Eliminated all static variables in the drivers.

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -43,74 +46,119 @@
+
+// AK8963, mag sensor address
+#define AK8963_MAG_I2C_ADDRESS          0x0C
+#define AK8963_Device_ID                0x48
```
Done.

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -43,74 +46,119 @@
+#define ST1_DATA_OVERRUN                0x02
+
+#define ST2_DATA_ERROR                  0x02
+#define ST2_MAG_SENSOR_OVERFLOW         0x03
```
Good catch. Something is completely wrong here. Likely an artifact from an former AK8975 driver. Fixed.

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -43,74 +46,119 @@
+#define CNTL1_MODE_CONT2                0x06
+#define CNTL1_MODE_SELF_TEST            0x08
+#define CNTL1_MODE_FUSE_ROM             0x0F
+#define CNTL1_BIT_14_Bit                0x00
```
Done.

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -43,74 +46,119 @@
+
+    ak8963SpiWriteRegisterDelay(bus, MPU_RA_I2C_SLV0_ADDR, slavedev->busdev_u.i2c.address | READ_FLAG); // set I2C slave address for read
+    ak8963SpiWriteRegisterDelay(bus, MPU_RA_I2C_SLV0_REG, reg);                     // set I2C slave register
+    ak8963SpiWriteRegisterDelay(bus, MPU_RA_I2C_SLV0_CTRL, len | 0x80);             // read number of bytes
```
Not commonly used in BF code yet. Keep.

ledvinap replied:
It's `I2C_SLV0_EN`. Using 0x80 directly means that register map has to be consulted ... 
Also `len` is only 4 bits, masking it before write (`len & 0x0f`) will document this ...

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -43,74 +46,119 @@
+
+    ak8963SpiWriteRegisterDelay(bus, MPU_RA_I2C_SLV0_ADDR, slavedev->busdev_u.i2c.address); // set I2C slave address for write
+    ak8963SpiWriteRegisterDelay(bus, MPU_RA_I2C_SLV0_REG, reg);                     // set I2C slave register
+    ak8963SpiWriteRegisterDelay(bus, MPU_RA_I2C_SLV0_DO, data);                     // set I2C salve value
```
Comes from the Gyro definitions and is directly derived from register names in the data sheet. Keep.

brucesdad13 replied:
I think he's just pointing out a small typo... 
![image](https://user-images.githubusercontent.com/23079971/31468789-5b83dbfa-aead-11e7-8fc0-2e312924f01e.png)

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -127,83 +175,53 @@ static uint32_t ak8963SensorQueuedReadTimeRemaining(void)
-    static ak8963ReadState_e state = CHECK_STATUS;
+    // we cannot use the ak8963ReadRegisterBuffer() method for SPI, it is to slow and blocks for far too long.
 
     bool retry = true;
```
Is used in WAIT_FOR_STATUS condition (23 lines below).

ledvinap replied:
Only way to enter `WAITING_FOR_STATUS` is through function call, with `retry == true`. Restart jump goes to `CHECK_STATUS` and it will always `return`.

BTW: simply returning with `state = CHECK_STATUS` and writing the request on next invocation will simplify code. And check failure should be rare, so wasted cycle is not a big problem

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -127,83 +175,53 @@ static uint32_t ak8963SensorQueuedReadTimeRemaining(void)
         case CHECK_STATUS:
-            ak8963SensorStartRead(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_ST1, 1);
+            ak8963SlaveStartRead(busdev, AK8963_MAG_REG_ST1, 1);
             state++;
```
Done.

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -219,33 +237,103 @@ static bool ak8963Read(int16_t *magData)
+}
+#endif
+
+static bool ak8963ReadRegisterBuffer(const busDevice_t *busdev, uint8_t reg, uint8_t len, uint8_t *buf)
```
Left for unchanged now.

MJ666 commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -219,33 +237,103 @@ static bool ak8963Read(int16_t *magData)
+}
+#endif
+
+static bool ak8963ReadRegisterBuffer(const busDevice_t *busdev, uint8_t reg, uint8_t len, uint8_t *buf)
```
This arguments get reordered some time ago. This is likely an leftover from this time. Fixed.

MJ666 commented:
Rebase an tested again. I did some more investigation regarding the MagGain of the AK8963 and the existing code is what is reasonable working so I stayed with this. I also added the same code to the AK8975 driver since it is basically the same.

ledvinap commented:
Calibration with `int8_t` makes no sense.
with signed:
ASA 0..127 maps to 50% .. 100% correcting factor
ASA 128-255 (-128..-1 when signed) maps to 0% .. 50% correcting factor.
There is no possibility to compensate value up (compensated value is always smaller than value in register) and there is calibration value (0x80) that results in null output independent of register reading. 

With unsigned:
ASA 0..255 maps to 50% .. 150%

Also datasheet does not mention anything about signedness (ones complement, twos complement,  sign bit), but specifies `COEFX7 COEFX6 COEFX5 COEFX4 COEFX3 COEFX2 COEFX1 COEFX0` + offset (-128) in equation.

MJ666 commented:
@ledvinap im a bit clueless here? Have tried quit a few things here with no promising results. How you would write the calculation here. Also the negative sign at the beginning worries me. But this is how it was working since a long time. Have you seen issues with iNav and this code. Very similar code is used a couple of times on the net.

ledvinap commented:
@MJ666 : Sorry, I don't quite understand ... 

Just use `uint8_t` as type entering equation from datasheet. It will give identical result for values 0x00 ... 0x7f and will scale register value up for 0x80..0xff.

The `*30` part is another nonsense. AK8963 sensitivity is 0.6uT/LSB, HMC5883l is set to 0.152uT/LSB(and calibrated), so it may make sense to multiply value by 4 to match range of both sensors. 
Or set BIT=1 - enable 16bit mode on AK8963, this will change sensitivity to 0.15uT/LSB (measurement range stays the same, it looks that result is simply shifted by 2 bits internally).

`* 30` also means that resulting value may overflow - reading is returned as `int16_t` and read value is +-8186 * 30. This won't be normally visible due to 65uT of earth's magnetic field is still within range.

For your chip calibration, (invalid) values calculated by existing equation are quite close to expected value of 4, (5.8, 6, 4.7). That may be reason it works for you.

IMO correct implementation is:
```
magGain[X] = ((((float)calibration[X] - 128) / 256) + 1) * 4;
```

Also calibration values may be stored multiplied by 256 as `int16_t`, saving some ram and cpu cycles
```
magGain[X] = (calibration[X] + 128) * 4;
magData[X] = -(int)(int16_t)(buf[1] << 8 | buf[0]) * magGain[X] / 256;
```

and for hmc5883l:
```
magGain[X] = abs(int(660.0 * HMC58X3_X_SELF_TEST_GAUSS * 2.0 * 10.0 * 256) / xyz_total[X]);
magData[X] = (int)(int16_t)(buf[0] << 8 | buf[1]) * magGain[X] / 256;

MJ666 commented:
Did some more updates. I updated the Mag gain calculation according to @ledvinap's recommendation. Tested the code with float and integer storage of the mag gains. Looks to be there is an error of about 20% between both variants. For now both code variants are in the code and the integer version is commented out. Mag sensor values look reasonable now. we need to decide wich way we want to go accuracy vs. speed.

ledvinap commented:
@mj666: that's very strange. Integer variant just moved division (by 256) into calibrated value calculation. Results shall be identical(or within 1lsb due to rounding)

MJ666 commented:
Rebased. Also got the modified NAZE target to work for testing the HMV5883L in I2C mode. Also the modified Mag interrupt is working. Actually the coded contains the former Float gain calculation and the modified integer calculation. The Integer version can be activated via defining INTEGER_MAG_GAIN variable during building (make TARGET="target_name" OPTIONS="INTEGER_MAG_GAIN". Both variants still give different results for some reason? The difference is about 10%. The general value of the measures are consistent between versions. I also find out the slave part of the AK8963 driver still using static definitions. I prefer resolve this in an future run. It is only compiled in for MPU9250 an MPU6500 in SPI mode when AK8963 driver is also used. Now only the HMV5883L in SPI mode need to be tested. We still need to review the Mag gain calculations and decide wich one should stay. The NAZE taget changes are also only for testing and need to be removed before merging.

MJ666 commented:
Found another issue the AK8963 and the AK8975 driver can be knocked out when you force an overflow with an strong magnet. Likely continuous mode will be better to be used. It will loop in this error situation on will not start a new cycle.

MJ666 commented:
Fixed the overflow problem. the slave state machine is now working reliable without lookups. Need to bring the retry goto construct back. Without this the state machine is looking up after a while.

MJ666 commented:
Looks to be there is no target using the HMC5883L in SPI mode. So we at not able to test by now. looks to be wee need finally check the gain calculation if it is correct and decide if we keep this in float or move the integer for this. Please give me your feedback here so I can finalize this PR. from my point of view is should be than ready to go. looking forward to your feedback. Thanks.

MJ666 commented:
The compass task can be reschedule only if the slave mode is really active independent of the Mag definitions. Without this change the compass task is always running with 40Hz as long MPU6500 or MPU9250 in SPI mode and AK8963 driver is configured. for AlienFlightNG F7 FC's we have hardware with MPU6500 SPI, MPU9250 SPI or ICM20602 and AK8963 SPI.

ledvinap commented on src/main/drivers/compass/compass_ak8963.c:
```diff
@@ -213,90 +231,227 @@ static bool ak8963Read(int16_t *magData)
                 if (retry) {
```
As I said before, this jump is always taken. Your previous version of code did honor `retry`, which may have caused lockups ...

MJ666 replied:
I prefer to not touch this for now. The state engine need some review again to remove last statics. This PR changes getting much more impact already now then getting the AK8963 SPI driver working.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -259,50 +315,63 @@ static bool hmc5883lInit(void)
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFA, 0x70);   // Configuration Register A  -- 0 11 100 00  num samples: 8 ; output rate: 15Hz ; normal measurement mode
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFB, 0x20);   // Configuration Register B  -- 001 00000    configuration gain 1.3Ga
-    i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_MODE, 0x00);    // Mode register             -- 000000 00    continuous Conversion Mode
+    magdev->magGain[X] = (int16_t)fabsf((660.0f * HMC58X3_X_SELF_TEST_GAUSS * 2.0f * 10.0f * 256.0f) / xyz_total[X]);
```
integer conversion must be performed last (or implicitly). As implemented, code will cast first (with overflow), then divide.

Also sign may be wrong - `fabsf()` is applied to positive constant, but `xyz_total` may be negative.

```
magdev->magGain[X] = ABS((int)(660.0f * HMC58X3_X_SELF_TEST_GAUSS * 2.0f * 10.0f * 256.0f) / xyz_total[X]);
```

MJ666 commented on src/main/drivers/bus.h:
```diff
@@ -33,15 +33,17 @@ typedef struct busDevice_s {
-#define BUSTYPE_NONE 0
-#define BUSTYPE_I2C  1
-#define BUSTYPE_SPI  2
+#define BUSTYPE_NONE    0
```
Changing this is causing some more changes to all dispatch functions. Here I removed the redundant dispatch functions also from the Baro drivers.

MJ666 commented:
The change of the bus driver had impact on all driver bus dispatch code. I removed now the redundant dispatch code from all mag and baro drivers. The BMP280 is tested in SPI and I2Cmode. The MS5611 is tested in I2C mode on the targets I have here. All former tests for the Mag driver are also repeated and everything worked as expected.
Finally, the mag gain calculations gave same results with integer and float calculation. Removed the float version from code.
I also restored the Naze configuration to the original settings.
In my view this is now well tested and ready to go.

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+    IOInit(magIntIO, OWNER_COMPASS_EXTI, 0);
+    IOConfigGPIO(magIntIO, IOCFG_IN_FLOATING);
+    EXTIHandlerInit(&mag->exti, hmc5883_extiHandler);
+    EXTIConfig(magIntIO, &mag->exti, NVIC_PRIO_MAG_INT_EXTI, EXTI_Trigger_Rising);
```
Just a note:
F1,F3,F4 EXTIConfig may be eventually updated to handle IO configuration internally (to match F7)

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+#if defined (STM32F7)
+    IOInit(magIntIO, OWNER_COMPASS_EXTI, 0);
+    EXTIHandlerInit(&mag->exti, hmc5883_extiHandler);
+    EXTIConfig(magIntIO, &gmag->exti, NVIC_PRIO_MPU_INT_EXTI, IO_CONFIG(GPIO_MODE_INPUT,0,GPIO_NOPULL));
```
If possible, please call `EXTIEnable(magIntIO, true);` here too

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+ * HMC5983 datasheet states SPI mode is
+ *     SCK is high when CS is high
+ *     Data is sampled at the rising edge of SCK
+ * However, it seems that SCK condition is ignored (works with SCK low when CS is high).
```
SPI slave should ignore SCK as long as CS is not asserted. 
Only SCK state when asserting CS may need some consideration

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+ */
+static void hmc5883SpiInit(busDevice_t *busdev)
+{   
+    static bool hardwareInitialised = false;
```
This will fail with multiple busdev (only first will be initialized)

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+}
 #endif
+
+static bool hmc5883lRead(magDev_t *magdev, int16_t *magData)
```
( `magDev_t *mag` , to match hmc5883lConfigureDataReadyInterruptHandling)

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+
+    hmc5883lRead(magdev, magADC);
 
     for (i = 0; i < 10; i++) {  // Collect 10 samples
```
There is big overlap in positive/negative measurement code. It would be nice to move it into separate function or maybe unify it using outer loop.

ledvinap replied:
Separate function would be better, just for example:
```
  enum {
    polPos, polNeg
  };
  for (int polarity = polPos; polarity <= polNeg; polarity++) {
  switch(polarity) {
  case polPos:
    busWriteRegister(busdev, HMC58X3_REG_CONFA, HMC_CONFA_DOR_15HZ | HMC_CONFA_POS_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to pos bias
    break;
  case polNeg:
    busWriteRegister(busdev, HMC58X3_REG_CONFA, HMC_CONFA_DOR_15HZ | HMC_CONFA_NEG_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to negative bias.
    break;
  }
  for(int i =0; i< 10; i++) ... 
  ...
  xyz_total[X] += ((polarity == polPos) ? 1 : -1) * magADC[X];

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -259,50 +303,51 @@ static bool hmc5883lInit(void)
-    magGain[X] = fabsf(660.0f * HMC58X3_X_SELF_TEST_GAUSS * 2.0f * 10.0f / xyz_total[X]);
-    magGain[Y] = fabsf(660.0f * HMC58X3_Y_SELF_TEST_GAUSS * 2.0f * 10.0f / xyz_total[Y]);
-    magGain[Z] = fabsf(660.0f * HMC58X3_Z_SELF_TEST_GAUSS * 2.0f * 10.0f / xyz_total[Z]);
+    magdev->magGain[X] = ABS((int32_t)(660.0f * HMC58X3_X_SELF_TEST_GAUSS * 2.0f * 10.0f * 256.0f) / xyz_total[X]);
```
BTW: is `ABS` necessary? Positive bias is added, negative subtracted, so result should be (2*) positive. the `fabs` looks like some old hack ... 

I would also prefer casting to `int`, there is no explicit need for 32 bits

ledvinap commented on src/main/drivers/compass/compass_hmc5883l.c:
```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
-    magData[X] = (int16_t)(buf[0] << 8 | buf[1]) * magGain[X];
-    magData[Z] = (int16_t)(buf[2] << 8 | buf[3]) * magGain[Z];
-    magData[Y] = (int16_t)(buf[4] << 8 | buf[5]) * magGain[Y];
+    magData[X] = (int16_t)(buf[0] << 8 | buf[1]) * magdev->magGain[X] / 256;
```
This may be problematic - gain may be larger than 1.0 (> 256), `int16_t` will then overflow on saturated magnetometer.
It is quite extreme case and reading is unusable anyway (earth's magnetic field is too small to overflow), but it is better to handle it anyway.
IMO it is easier/enough to use saturation to `int16_t` range:

```
static int16_t parseMag(uint8* raw, int16_t gain) {
  int ret = (int16_t)(raw[0] << 8 | raw[1]) * gain / 256;
  return constrain(ret, INT16_MIN, INT16_MAX);
}

...
magData[X] = parseMag(buf + 0, magdev->magGain[X]);
magData[Z] = parseMag(buf + 2, magdev->magGain[Z]);
magData[Y] = parseMag(buf + 4, magdev->magGain[Y]);

ledvinap replied:
(and the problem is the same for other magnetometers)

mikeller commented:
@MJ666: If you are done doing touch ups as per @ledvinap's remarks this can be merged.

MJ666 commented:
I'm traveling this week and will likely take care on this after my return. At least my testing posibility es are quite limited during my absence.

mikeller commented:
All good, there's still plenty of time left until 3.3. :-)

MJ666 commented:
Have implemented all recommended updates. This code is only tested with the AFNG F7 development board yet. The Mag sensor values are different to the ones before the change. i can do the remaining tests by end of the week. The original intention for this PR was to make the AK8963 SPI Mag work for the AlienFlightNG F7 Anubis16. I was not expecting to make so many changes to all of the Baro and Mag drivers. I was hoping to get the Mag for the Anubis16 working faster than in a couple of months from now. From 3.1.0 to 3.2.0 it tock us more than 8 months.

mikeller commented on src/main/drivers/bus.c:
```diff
@@ -26,33 +26,52 @@
-    }
-    return false;
+#if !defined(USE_SPI) && !defined(USE_I2C)
+    UNUSED(reg);
```
`busdev` is unused in this case too.

MJ666 replied:
Fixed.

mikeller commented:
@MJ666: Merged this now. If testing shows that fixes are needed, this can be done in a new pull request.
